bcgov/cas-registration
# Feat : Auth for idir and bceid

Addresses [Ticket 14](https://app.zenhub.com/workspaces/team-moose-sprint-board-60d216175a2689000ec774d0/issues/gh/bcgov/cas-registration/14)

## 🚧 Changes

    - Add NextAuth package
    - Add client/.env.local 
    - Update client/.env
    - Add client/app/api/auth/[...nextauth]/route.ts: retrieves and augments JWT
    - Add client/app/api/auth/logout/route.ts
    - Add client/app/types/next-auth.d.ts: allows defining shared interfaces
    - Update client/middlewares/stackMiddlewares.tsx
    - Update client/app/(authenticated)/dashboard/page.tsx
    - Update client/app/(onboarding)/home/page.tsx
    - Update docs/developer-guide.md

## 🚀 Impact:
   - Implementation of IDIR and BCeid authentication using NextAuth.js configured with Keycloak provider JWT with user session information.
   - User session information is augmented and propagated to RSC and client components using getServerSession
   - User session information is used to secure app routes and api calls.


## 📝 Notes:
 Add client/app/api/auth/[...nextauth]/route.ts has hard coded role "admin" which will be addressed in [265](https://app.zenhub.com/workspaces/team-moose-sprint-board-60d216175a2689000ec774d0/issues/gh/bcgov/cas-registration/265)

**NextAuth user session information**: When calling from the server-side i.e. in Route Handlers, React Server Components, API routes getServerSession requires passing the same object you would pass to NextAuth when initializing NextAuth.js. To do so, you can export your NextAuth.js options in [...nextauth].ts then import NextAuth.js options in NextAuth.js options in Route Handlers, React Server Components, API routes. See "idir_user_guid: localhost:3000/dashboard/profile"


##  🔬 Testing:

1. As per docs/developer-guide.md, configure client/.env and client/.env.local files using [1Password](https://climateactionsecretariat.1password.ca/) documents `OBPS FE env` and : `OBPS FE env.local` 
2.  From terminal command, install package libraries:
   ```
cd client && yarn 
   ```
3.  From terminal command, start the development server:
   ```
cd client && yarn dev
   ```
4. In your web browser, navigate to `http://localhost:3000`
5. You should be directed to the anonymous landing page
![image](https://github.com/bcgov/cas-registration/assets/120038448/bd8166b1-4f19-4993-82dc-0b3340d625e7)
6. Click "Program Administrator Log In"
7. You should be directed to IDIR login screen
![image](https://github.com/bcgov/cas-registration/assets/120038448/2e29ad50-559e-4c51-a756-c7953b9e17a8)
8. Login using your IDIR account
9. You should be redirected to the admin dashboard
![image](https://github.com/bcgov/cas-registration/assets/120038448/a0175c46-dd72-45d3-af42-c72c592bcdfe)
10. Right hand header should display your IDIR name and a "Log out" button
11. Click "Log out"
12. You should be directed to the anonymous landing page
![image](https://github.com/bcgov/cas-registration/assets/120038448/bd8166b1-4f19-4993-82dc-0b3340d625e7)
13. Repeat steps 6-12
14. You should have the same results which ensures NextAuth and IDIR federated sign in and sign out work as expected
15. Click you user name in the header
16. `http://localhost:3000/dashboard/profile` route should display with information from your Auth JWT
17. Optional if you have BCeid login:
18. Click "Industrial  Operator Log In"
19. You should be directed to BCeid login screen
![image](https://github.com/bcgov/cas-registration/assets/120038448/8c6ba73e-e9bc-4f31-97bd-8d5bf65f01f0)
20. Login in with a BCeid 
21. You should be redirected to the admin dashboard
